### ISSUE TITLE
Planned inconsistencies in binary distribution regarding `libxml2`

### ISSUE THREAD

**Issue by xtkoba** [termux/termux-packages#17202]
Similar to #17129.

SONAME change in `libxml2` package is known to break revdeps. Example:

```
$ readelf
CANNOT LINK EXECUTABLE "readelf": cannot find "libxml2.so" from verneed[5] in DT_NEEDED list for "/data/data/com.termux/files/usr/lib/libLLVM-16.so"
```

Currently SONAME conflict of `libxml2.so` does not seem to cause any real troubles, so I hesitate to carry on the SONAME change for now, because it will break a wide range of packages for a considerable amount of time.

Infra team, do you have any ideas? I think the breakage will actually be avoidable with your cooperation.

**Comment by xtkoba** [comment]
I think I have found a way to minimize the troubles. I'm going to abuse the package management system, so there may be some side effect.

The plan is as follows. SOVERSION suffix will be enabled in `libxml2` version `2.11.4-1` which will become a broken package that cannot be installed on device. Every revdep will temporarily depend on `libxml2 (>= 2.11.4-2)` but will be rebuilt against `libxml2 (= 2.11.4-1)`. After every revdep is rebuilt and uploaded, `libxml2` will be revbumped to `2.11.4-2` which will not be broken anymore.

This work will begin at July 2 0:00 UTC. After that, it is highly recommended that you refrain from upgrading any packages until `libxml2` version `2.11.4-2` is released.

**Comment by xtkoba** [comment]
To Python users:

Some Python modules, e.g. `lxml`, depend on `libxml2`. Unfortunately there is no way to rebuild them automatically, and they will be left broken until manually rebuilt.

To fix manually for `lxml`, please execute the following command after you install `libxml2 (>= 2.11.4-2)`:

```bash
pip install --force lxml
```

The same goes for other modules. Sorry for your inconvenience.

**Comment by Gallium70** [comment]
Unfortunately I got the narrow timing to force-install this [break-the-shell version](https://github.com/termux/termux-packages/commit/c1bde8087865e65fa0628838dcab65e35e1203bd) of libxml2 :(

Since `bin/sh` is already replaced by the script itself, calling `/bin/sh` inside it actually cause an infinite loop calling itself. `dpkg` then is broken, because many scripts are using `/bin/sh`. I had to manually edit them and replace with `/bin/bash` and reinstall `dash` to regain `/bin/sh`.

The output of `ps aux | grep dpkg`:

![image](https://github.com/termux/termux-packages/assets/52118815/03fd8efb-754f-4b4e-b238-08e0ee9014b2)

Maybe there's a less destructive way to do this, e.g. calling `/bin/bash` instead of itself and add an `echo 'warning something'` ?

**Comment by xtkoba** [comment]
Now I am waiting for <a href="https://github.com/termux/termux-packages/actions/runs/5434726259">the rebuild of `libllvm`</a> to be finished. This will take several more hours. After that, PR #17302 will be merged and the bindist will be unbroken.

**Comment by xtkoba** [comment]
@Gallium70 Apologies for that. I'm not going to improve it anymore though.

**Comment by Gallium70** [comment]
@xtkoba Yeah, it's mostly my fault that I shouldn't play with `dpkg --force-overwrite'. Didn't expect that one day I would become an example of 'fool-proof can't proof a forceful fool'. 🤣

**Comment by ascendbeing** [comment]
The mephi.ru tmux mirror isn't up to date with the broken packages thankfully. I was really confused until I thought of trying Russian and Chinese mirrors and that one specifically is reachable and has a non broken libxml2

**Comment by xtkoba** [comment]
Almost all revdeps have been rebuilt and uploaded to the main server.

The `zig` package is an exception, because it cannot be rebuilt. We should consider disabling (removing) this package.

If you still see issues after a while, then there are possibilities that the dependencies in our packages are not specified correctly. In that case please file individual issues for them, instead of reporting in the form of comments in this issue.

I'm going to keep this issue open for a while, because theoretically the Python module issue persists indefinitely.

**Comment by xtkoba** [comment]
The workaround for Python modules in https://github.com/termux/termux-packages/issues/17202#issuecomment-1616385953 is not correct as of now. I'm trying to revise it.

**Comment by xtkoba** [comment]
Not included in the title but obviously this also breaks third-party packages that depend on the shared library in our `libxml2` package.

Note that this breakage is not related to upstream (libxml2 project) at all. Just to be sure.

**Comment by finagolfin** [comment]
This can be closed now?

**Comment by Revnoplex** [comment]
I suddenly can't run ffmpeg as I am running into the following error
```CANNOT LINK EXECUTABLE "ffmpeg": cannot find "libxml2.so" from verneed[0] in DT_NEEDED list for "/data/data/com.termux/files/usr/lib/libbluray.so"
```
Is this related to this issue?

**Comment by Revnoplex** [comment]
> @Revnoplex 
> patchelf --replace-needed libxml2.so  libxml2.so.2  /data/data/com.termux/files/usr/lib/libbluray.so

It worked, thanks!

**Comment by 9017232799** [comment]
Termux

**Comment by vstavrinov** [comment]
`ImportError: dlopen failed: cannot find "libxml2.so" from verneed[1] in DT_NEEDED list for "/data/data/com.termux/files/usr/lib/python3.11/site-packages/lxml/etree.cpython-311.so"`

The problem still exists

**Comment by romanovj** [comment]
@vstavrinov 
```
patchelf --print-needed /data/data/com.termux/files/usr/lib/python3.11/site-packages/lxml/etree.cpython-311.so
libandroid-support.so
libxslt.so
libexslt.so
libxml2.so.2
libc.so
libpython3.11.so.1.0
```
